controller/program/tests/tests/set_sol_value_calculator.rs

```diff
@@ -0,0 +1,584 @@
+
+    assert_ne!(
+        lst_state_bef.sol_value_calculator,
+        lst_state_aft.sol_value_calculator
+    );
+    assert_eq!(&lst_state_aft.sol_value_calculator, expected_new_calc);
+
+    let delta = i128::from(pool_aft.total_sol_value) - i128::from(pool_bef.total_sol_value);
+    assert_eq!(
+        delta,
+        i128::from(lst_state_aft.sol_value) - i128::from(lst_state_bef.sol_value)
+    );
+}
+
+#[test]
+fn set_sol_value_calculator_jupsol_fixture() {
+    let pool_pk = Pubkey::new_from_array(POOL_STATE_ID);
+    let pool_acc = ALL_FIXTURES
+        .get(&pool_pk)
+        .expect("missing pool state fixture");
+
+    let pool = PoolStatePacked::of_acc_data(&pool_acc.data)
+        .unwrap()
+        .into_pool_state();
+    let admin = pool.admin;
+
+    let ix_prefix = set_sol_value_calculator_ix_pre_keys_owned(
+        admin,
+        &TOKENKEG_PROGRAM,
+        JUPSOL_MINT.to_bytes(),
+    );
+    let builder = SetSolValueCalculatorKeysBuilder {
+        ix_prefix,
+        calc_prog: *SvcAgTy::SanctumSplMulti(()).svc_program_id(),
+        calc: jupsol_fixtures_svc_suf(),
+    };
+    let ix = set_sol_value_calculator_ix(&builder, JUPSOL_FIXTURE_LST_IDX as u32);
+    let mut accounts = set_sol_value_calculator_fixtures_accounts_opt(&builder);
+
+    upsert_account(
+        &mut accounts,
+        (
+            Pubkey::new_from_array(admin),
+            Account {
+                lamports: u64::MAX,
+                ..Default::default()
+            },
+        ),
+    );
+
+    let lsl_pk = Pubkey::new_from_array(LST_STATE_LIST_ID);
+    let lsl_acc = ALL_FIXTURES.get(&lsl_pk).unwrap().clone();
+    let mut lsl_data = lsl_acc.data.to_vec();
+
+    // Set initial calculator to a random pubkey
+    let lsl_mut = LstStatePackedListMut::of_acc_data(&mut lsl_data).unwrap();
+    let lst_mut = unsafe {
+        lsl_mut
+            .0
+            .get_mut(JUPSOL_FIXTURE_LST_IDX)
+            .unwrap()
+            .as_lst_state_mut()
+    };
+    lst_mut.sol_value_calculator = Pubkey::new_unique().to_bytes();
+
+    upsert_account(&mut accounts, (lsl_pk, lst_state_list_account(lsl_data)));
+
+    let InstructionResult {
+        program_result,
+        resulting_accounts,
+        ..
+    } = SVM.with(|svm| svm.process_instruction(&ix, &accounts));
+
+    assert_eq!(program_result, ProgramResult::Success);
+
+    assert_correct_set(
+        &accounts,
+        &resulting_accounts,
+        JUPSOL_MINT.as_array(),
+        &jupsol_fixtures_svc_suf().svc_program_id(),
+    );
+}
+
+enum TestErrorType {
+    Unauthorized,
+    PoolRebalancing,
+    PoolDisabled,
+}
+
+fn set_sol_value_calculator_proptest(
+    pool: PoolState,
+    mut lsl: LstStateListData,
+    lsd: LstStateData,
+    admin: [u8; 32],
+    calc_prog: [u8; 32],
+    calc: SvcCalcAccsAg,
+    initial_calc_prog: [u8; 32],
+    new_balance: u64,
+    additional_upserts: impl FnOnce(&mut Vec<PkAccountTup>, &LstStateData),
```

Comment:
can this just be a `impl IntoIterator<Item = PkAccountTup>` (like a `Vec<PkAccountTup>`) instead of a closure?